File: cadwiki-nuget/cadwiki.NUnitTestRunner/Creators/TestEvidenceCreator.vb
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.IO
Imports System.Windows.Automation
Imports System.Windows.Automation.AutomationElement
Imports cadwiki.NUnitTestRunner.Results
Imports cadwiki.NUnitTestRunner.TestEvidence
Imports Microsoft.Test.Input
Imports Image = cadwiki.NUnitTestRunner.TestEvidence.Image

Namespace Creators

    Public Class TestEvidenceCreator

        Private Shared _evidenceForCurrentlyExecutingTest As Evidence
        Private Shared _localFolderCache As String = IO.Path.GetTempPath + "cadwiki.NUnitTestRunner"
        Private Shared _localScreenShotCache As String = _localFolderCache + "\" + "screenshots"
        Private Shared _pdfFileReport As String = "AutomatedTestEvidence.pdf"
        Private Shared _jsonFileResults As String = "AutomatedTestEvidence.json"

        Public Sub New()
            If (_evidenceForCurrentlyExecutingTest Is Nothing) Then
                _evidenceForCurrentlyExecutingTest = New Evidence()
            End If
            If (Not IO.Directory.Exists(_localFolderCache)) Then
                IO.Directory.CreateDirectory(_localFolderCache)
            End If
            If (Not IO.Directory.Exists(_localScreenShotCache)) Then
                IO.Directory.CreateDirectory(_localScreenShotCache)
            End If
        End Sub

        Public Function CreatePdf(suiteResult As ObservableTestSuiteResults) As String
            Dim pdfCreator As New PdfCreator(GetNewPdfReportFilePath(suiteResult))
            pdfCreator.AddTitlePage(suiteResult)
            For Each testResult As TestResult In suiteResult.TestResults
                pdfCreator.AddTestPage(testResult)
            Next
            pdfCreator.Save()
            Return pdfCreator.PdfFilePath
        End Function

        Public Function WriteTestSuiteResultsToFile(suiteResult As ObservableTestSuiteResults, jsonString As String) As String
            Dim jsonFilePath As String
            If (Not String.IsNullOrEmpty(suiteResult.TestSuiteName)) Then
                jsonFilePath = _localFolderCache + "\" + suiteResult.TestSuiteName + "-" + _jsonFileResults
            Else
                jsonFilePath = _localFolderCache + "\" + _jsonFileResults
            End If
            Dim jsonFile As String = NetUtils.Paths.GetUniqueFilePath(jsonFilePath)
            File.WriteAllText(jsonFile, jsonString)
            Return jsonFile
        End Function

        Public Function GetNewPdfReportFilePath(suiteResult As ObservableTestSuiteResults) As String
            Dim reportFilePath As String
            If (Not String.IsNullOrEmpty(suiteResult.TestSuiteName)) Then
                reportFilePath = _localFolderCache + "\" + suiteResult.TestSuiteName + "-" + _pdfFileReport
            Else
                reportFilePath = _localFolderCache + "\" + _pdfFileReport
            End If
            reportFilePath = NetUtils.Paths.GetUniqueFilePath(reportFilePath)
            Return reportFilePath
        End Function

        Public Function GetFolderCache() As String
            Return _localFolderCache
        End Function

        Public Function GetScreenshotCache() As String
            Return _localScreenShotCache
        End Function

        Public Sub SetEvidenceForCurrentTest(testEvidence As Evidence)
            _evidenceForCurrentlyExecutingTest = testEvidence
        End Sub

        Public Function GetEvidenceForCurrentTest() As Evidence
            Return _evidenceForCurrentlyExecutingTest
        End Function

        Public Function ProcessesGetHandleFromUiTitle(ByVal windowTitle As String) As IntPtr
            Dim hWnd As IntPtr = IntPtr.Zero

            For Each pList As Process In Process.GetProcesses()

                If pList.MainWindowTitle.Contains(windowTitle) Then
                    hWnd = pList.MainWindowHandle
                End If
            Next

            Return hWnd
        End Function

        Public Shared Sub PrintWindowToImage(ByVal windowIntPtr As IntPtr, ByVal screenshotPath As String, ByVal format As ImageFormat)
            Dim screenshot As Bitmap = PrintWindowWithWinAPI(windowIntPtr)
            screenshot.Save(screenshotPath, format)
        End Sub


        Public Function MicrosoftTestClickUiControl(windowIntPtr As IntPtr, automationId As String) As Boolean
            Return MicrosoftTestClickUiControlByAutomationId(windowIntPtr, automationId)
        End Function

        Public Function MicrosoftTestClickUiControlByAutomationId(windowIntPtr As IntPtr, automationId As String) As Boolean

            Dim element As AutomationElement = GetElementByAutomationId(windowIntPtr, automationId)
            If (element Is Nothing) Then
                Return False
            End If
            Dim clickableSystemDrawingPoint As System.Drawing.Point =
                GetClickableSystemDrawingPointFromElement(element)

            Return MicrosoftTestClickPoint(clickableSystemDrawingPoint)
        End Function

        Public Function MicrosoftTestClickUiControlByName(windowIntPtr As IntPtr, name As String) As Boolean

            Dim element As AutomationElement = GetElementByControlName(windowIntPtr, name)
            If (element Is Nothing) Then
                Return False
            End If
            Dim clickableSystemDrawingPoint As System.Drawing.Point =
                GetClickableSystemDrawingPointFromElement(element)

            Return MicrosoftTestClickPoint(clickableSystemDrawingPoint)
        End Function

        Public Function GetControlCollection(windowIntPtr As IntPtr) As AutomationElementCollection
            Dim root = AutomationElement.FromHandle(windowIntPtr)
            Dim elementCollection As AutomationElementCollection = root.FindAll(TreeScope.Subtree, System.Windows.Automation.Condition.TrueCondition)
            Return elementCollection
        End Function

        Public Sub TakeJpegScreenshot(windowIntPtr As IntPtr, title As String)
            Dim fileName As String = title + ".jpg"
            fileName = NetUtils.Paths.ReplaceAllillegalCharsForWindowsOSInFileName(fileName, "-")
            Dim screenshotPath As String = GetScreenshotCache() + "\" + fileName
            screenshotPath = NetUtils.Paths.GetUniqueFilePath(screenshotPath)
            Dim format As ImageFormat = ImageFormat.Jpeg
            TestEvidenceCreator.PrintWindowToImage(windowIntPtr, screenshotPath, format)
            Dim image As New Image()
            image.Title = title
            image.FilePath = screenshotPath
            _evidenceForCurrentlyExecutingTest.Images.Add(image)
            Me.SetEvidenceForCurrentTest(_evidenceForCurrentlyExecutingTest)
        End Sub

        Private Shared Function PrintWindowWithWinAPI(ByVal hwnd As IntPtr) As Bitmap
            Dim rc As WinAPI.RECT
            WinAPI.Stubs.GetWindowRect(hwnd, rc)
            Dim bmp As Bitmap = New Bitmap(rc.Width, rc.Height, PixelFormat.Format32bppArgb)
            Dim gfxBmp As Graphics = Graphics.FromImage(bmp)
            Dim hdcBitmap As IntPtr = gfxBmp.GetHdc()
            WinAPI.Stubs.PrintWindow(hwnd, hdcBitmap, 0)
            gfxBmp.ReleaseHdc(hdcBitmap)
            gfxBmp.Dispose()
            Return bmp
        End Function



        Private Function GetElementByAutomationId(ByVal windowIntPtr As IntPtr, ByVal automationIdToFind As String) As AutomationElement
            Dim root = AutomationElement.FromHandle(windowIntPtr)
            Dim elementCollection As AutomationElementCollection = root.FindAll(TreeScope.Subtree, System.Windows.Automation.Condition.TrueCondition)

            For Each element As AutomationElement In elementCollection
                Dim current As AutomationElementInformation = element.Current
                Dim controlType As ControlType = current.ControlType
                Dim name As String = current.Name
                Dim automationId As String = current.AutomationId
                If automationId.Equals(automationIdToFind) Then
                    Return element
                End If
            Next

            Return Nothing
        End Function

        Private Function GetElementByControlName(ByVal windowIntPtr As IntPtr, ByVal controlNameToFind As String) As AutomationElement
            Dim root = AutomationElement.FromHandle(windowIntPtr)
            Dim elementCollection As AutomationElementCollection = root.FindAll(TreeScope.Subtree, System.Windows.Automation.Condition.TrueCondition)

            For Each element As AutomationElement In elementCollection
                Dim current As AutomationElementInformation = element.Current
                Dim controlType As ControlType = current.ControlType
                Dim name As String = current.Name
                Dim automationId As String = current.AutomationId
                If name.Equals(controlNameToFind) Then
                    Return element
                End If
            Next

            Return Nothing
        End Function

        Private Function GetClickableSystemDrawingPointFromElement(ByVal element As AutomationElement) As System.Drawing.Point
            Dim windowsPoint As System.Windows.Point = element.GetClickablePoint()
            Dim drawingPoint As System.Drawing.Point = New System.Drawing.Point(CInt(windowsPoint.X), CInt(windowsPoint.Y))
            Return drawingPoint
        End Function

        Private Function MicrosoftTestClickPoint(ByVal clickableSystemDrawingPoint As System.Drawing.Point) As Boolean
            Microsoft.Test.Input.Mouse.MoveTo(clickableSystemDrawingPoint)
            Microsoft.Test.Input.Mouse.Click(MouseButton.Left)
            Return True
        End Function

    End Class












End Namespace

File: cadwiki-nuget/cadwiki.NUnitTestRunner/Creators/TestEvidenceCreator.cs
using System;
using System.Diagnostics;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Windows.Automation;
using static System.Windows.Automation.AutomationElement;
using cadwiki.NUnitTestRunner.Results;
using cadwiki.NUnitTestRunner.TestEvidence;
using Image = cadwiki.NUnitTestRunner.TestEvidence.Image;
using Microsoft.Test.Input;

namespace cadwiki.NUnitTestRunner.Creators
{

    public class TestEvidenceCreator
    {

        private static Evidence _evidenceForCurrentlyExecutingTest;
        private static string _localFolderCache = Path.GetTempPath() + "cadwiki.NUnitTestRunner";
        private static string _localScreenShotCache = _localFolderCache + @"\" + "screenshots";
        private static string _pdfFileReport = "AutomatedTestEvidence.pdf";
        private static string _jsonFileResults = "AutomatedTestEvidence.json";
        private static string _htmlFileReport = "AutomatedTestEvidence.html";

        public TestEvidenceCreator()
        {
            if (_evidenceForCurrentlyExecutingTest is null)
            {
                _evidenceForCurrentlyExecutingTest = new Evidence();
            }
            if (!Directory.Exists(_localFolderCache))
            {
                Directory.CreateDirectory(_localFolderCache);
            }
            if (!Directory.Exists(_localScreenShotCache))
            {
                Directory.CreateDirectory(_localScreenShotCache);
            }
        }

        public string CreatePdf(ObservableTestSuiteResults suiteResult)
        {
            var pdfCreator = new PdfCreator(GetNewPdfReportFilePath(suiteResult));
            pdfCreator.AddTitlePage(suiteResult);
            foreach (TestResult testResult in suiteResult.TestResults)
                pdfCreator.AddTestPage(testResult);
            pdfCreator.Save();
            return pdfCreator.PdfFilePath;
        }

        public void CreateHtmlReport(ObservableTestSuiteResults suiteResult)
        {
            var htmlCreator = new HtmlCreator();
            string htmlReportFilePath = GetNewHtmlReportFilePath(suiteResult);
            string reportFolder = Path.GetDirectoryName(htmlReportFilePath);

            var model = new HTML_Models.TestSuiteReport(suiteResult);
            model.TestSuiteResults.SetImagePathsToRelative(reportFolder);
            htmlCreator.ParameterizeReportTemplate(model);
            htmlCreator.SaveReportToFile(htmlReportFilePath);
        }

        public string WriteTestSuiteResultsToFile(ObservableTestSuiteResults suiteResult, string jsonString)
        {
            string jsonFilePath;
            if (!string.IsNullOrEmpty(suiteResult.TestSuiteName))
            {
                jsonFilePath = _localFolderCache + @"\" + suiteResult.TestSuiteName + "-" + _jsonFileResults;
            }
            else
            {
                jsonFilePath = _localFolderCache + @"\" + _jsonFileResults;
            }
            string jsonFile = NetUtils.Paths.GetUniqueFilePath(jsonFilePath);
            File.WriteAllText(jsonFile, jsonString);
            return jsonFile;
        }

        public string GetNewPdfReportFilePath(ObservableTestSuiteResults suiteResult)
        {
            string reportFilePath;
            if (!string.IsNullOrEmpty(suiteResult.TestSuiteName))
            {
                reportFilePath = _localFolderCache + @"\" + suiteResult.TestSuiteName + "-" + _pdfFileReport;
            }
            else
            {
                reportFilePath = _localFolderCache + @"\" + _pdfFileReport;
            }
            reportFilePath = NetUtils.Paths.GetUniqueFilePath(reportFilePath);
            return reportFilePath;
        }

        public string GetNewHtmlReportFilePath(ObservableTestSuiteResults suiteResult)
        {
            string reportFilePath;
            if (!string.IsNullOrEmpty(suiteResult.TestSuiteName))
            {
                reportFilePath = _localFolderCache + @"\" + suiteResult.TestSuiteName + "-" + _htmlFileReport;
            }
            else
            {
                reportFilePath = _localFolderCache + @"\" + _htmlFileReport;
            }
            reportFilePath = NetUtils.Paths.GetUniqueFilePath(reportFilePath);
            return reportFilePath;
        }

        public string GetFolderCache()
        {
            return _localFolderCache;
        }

        public string GetScreenshotCache()
        {
            return _localScreenShotCache;
        }

        public void SetEvidenceForCurrentTest(Evidence testEvidence)
        {
            _evidenceForCurrentlyExecutingTest = testEvidence;
        }

        public Evidence GetEvidenceForCurrentTest()
        {
            return _evidenceForCurrentlyExecutingTest;
        }

        public IntPtr ProcessesGetHandleFromUiTitle(string windowTitle)
        {
            var hWnd = IntPtr.Zero;

            foreach (Process pList in Process.GetProcesses())
            {

                if (pList.MainWindowTitle.Contains(windowTitle))
                {
                    hWnd = pList.MainWindowHandle;
                }
            }

            return hWnd;
        }

        public static void PrintWindowToImage(IntPtr windowIntPtr, string screenshotPath, ImageFormat format)
        {
            var screenshot = PrintWindowWithWinAPI(windowIntPtr);
            screenshot.Save(screenshotPath, format);
        }


        public bool MicrosoftTestClickUiControl(IntPtr windowIntPtr, string automationId)
        {
            return MicrosoftTestClickUiControlByAutomationId(windowIntPtr, automationId);
        }

        public bool MicrosoftTestClickUiControlByAutomationId(IntPtr windowIntPtr, string automationId)
        {

            var element = GetElementByAutomationId(windowIntPtr, automationId);
            if (element is null)
            {
                return false;
            }
            var clickableSystemDrawingPoint = GetClickableSystemDrawingPointFromElement(element);

            return MicrosoftTestClickPoint(clickableSystemDrawingPoint);
        }

        public bool MicrosoftTestClickUiControlByName(IntPtr windowIntPtr, string name)
        {

            var element = GetElementByControlName(windowIntPtr, name);
            if (element is null)
            {
                return false;
            }
            var clickableSystemDrawingPoint = GetClickableSystemDrawingPointFromElement(element);

            return MicrosoftTestClickPoint(clickableSystemDrawingPoint);
        }

        public AutomationElementCollection GetControlCollection(IntPtr windowIntPtr)
        {
            var root = FromHandle(windowIntPtr);
            var elementCollection = root.FindAll(TreeScope.Subtree, Condition.TrueCondition);
            return elementCollection;
        }

        public void TakeJpegScreenshot(IntPtr windowIntPtr, string title)
        {
            string fileName = title + ".jpg";
            fileName = NetUtils.Paths.ReplaceAllillegalCharsForWindowsOSInFileName(fileName, "-");
            string screenshotPath = GetScreenshotCache() + @"\" + fileName;
            screenshotPath = NetUtils.Paths.GetUniqueFilePath(screenshotPath);
            var format = ImageFormat.Jpeg;
            PrintWindowToImage(windowIntPtr, screenshotPath, format);
            var image = new Image();
            image.Title = title;
            image.FilePath = screenshotPath;
            _evidenceForCurrentlyExecutingTest.Images.Add(image);
            SetEvidenceForCurrentTest(_evidenceForCurrentlyExecutingTest);
        }

        private static Bitmap PrintWindowWithWinAPI(IntPtr hwnd)
        {
            WinAPI.RECT rc;
            WinAPI.Stubs.GetWindowRect(hwnd, out rc);
            var bmp = new Bitmap(rc.Width, rc.Height, PixelFormat.Format32bppArgb);
            var gfxBmp = Graphics.FromImage(bmp);
            var hdcBitmap = gfxBmp.GetHdc();
            WinAPI.Stubs.PrintWindow(hwnd, hdcBitmap, 0);
            gfxBmp.ReleaseHdc(hdcBitmap);
            gfxBmp.Dispose();
            return bmp;
        }



        private AutomationElement GetElementByAutomationId(IntPtr windowIntPtr, string automationIdToFind)
        {
            var root = FromHandle(windowIntPtr);
            var elementCollection = root.FindAll(TreeScope.Subtree, Condition.TrueCondition);

            foreach (AutomationElement element in elementCollection)
            {
                var current = element.Current;
                var controlType = current.ControlType;
                string name = current.Name;
                string automationId = current.AutomationId;
                if (automationId.Equals(automationIdToFind))
                {
                    return element;
                }
            }

            return null;
        }

        private AutomationElement GetElementByControlName(IntPtr windowIntPtr, string controlNameToFind)
        {
            var root = FromHandle(windowIntPtr);
            var elementCollection = root.FindAll(TreeScope.Subtree, Condition.TrueCondition);

            foreach (AutomationElement element in elementCollection)
            {
                var current = element.Current;
                var controlType = current.ControlType;
                string name = current.Name;
                string automationId = current.AutomationId;
                if (name.Equals(controlNameToFind))
                {
                    return element;
                }
            }

            return null;
        }

        private Point GetClickableSystemDrawingPointFromElement(AutomationElement element)
        {
            var windowsPoint = element.GetClickablePoint();
            var drawingPoint = new Point((int)Math.Round(windowsPoint.X), (int)Math.Round(windowsPoint.Y));
            return drawingPoint;
        }

        private bool MicrosoftTestClickPoint(Point clickableSystemDrawingPoint)
        {
            Mouse.MoveTo(clickableSystemDrawingPoint);
            Mouse.Click(MouseButton.Left);
            return true;
        }

    }












}